sesman/verify_user_pam.c

```diff
@@ -53,43 +54,117 @@ struct t_auth_info
     pam_handle_t *ph;
 };
 
+/***************************************************************************//**
+ * Returns a string representing a pam_conv message style
+ *
+ * @param msg_style PAM msg_style (pam_conv(3))
+ * @param buff      Buffer for conversion of unrecognised values
+ * @param bufflen   Total length of above
+ *
+ * The buffer described by buff is only written to if required.
+ */
+static const char *
+msg_style_to_str(int msg_style, char *buff, unsigned int bufflen)
+{
+    const char *result;
+    switch (msg_style)
+    {
+        case PAM_PROMPT_ECHO_OFF:
+            result = "PAM_PROMPT_ECHO_OFF";
+            break;
+
+        case PAM_PROMPT_ECHO_ON:
+            result = "PAM_PROMPT_ECHO_ON";
+            break;
+
+        case PAM_ERROR_MSG:
+            result = "PAM_ERROR_MSG";
+            break;
+
+        case PAM_TEXT_INFO:
+            result = "PAM_TEXT_INFO";
+            break;
+
+        default:
+            snprintf(buff, bufflen, "UNKNOWN_0x%x", msg_style);
+            result = buff;
+    }
+
+    return result;
+}
+
 /******************************************************************************/
 static int
 verify_pam_conv(int num_msg, const struct pam_message **msg,
                 struct pam_response **resp, void *appdata_ptr)
 {
     int i;
-    struct pam_response *reply;
+    struct pam_response *reply = NULL;
     struct t_user_pass *user_pass;
+    char sb[64];
+    int rv = PAM_SUCCESS;
+
+    if (num_msg <= 0 || num_msg > PAM_MAX_NUM_MSG)
+    {
+        rv = PAM_CONV_ERR;
+    }
+    else if ((reply = g_new0(struct pam_response, num_msg)) == NULL)
+    {
+        rv = PAM_BUF_ERR;
+    }
+    else
+    {
+        for (i = 0; i < num_msg && rv == PAM_SUCCESS; i++)
+        {
+            LOG_DEVEL(LOG_LEVEL_INFO, "Handling struct pam_message"
+                      " { style = %s, msg = \"%s\" }",
+                      msg_style_to_str(msg[i]->msg_style, sb, sizeof (sb)),
+                      msg[i]->msg == NULL ? "<null>" : msg[i]->msg);
 
-    reply = g_new0(struct pam_response, num_msg);
+            switch (msg[i]->msg_style)
+            {
+                case PAM_PROMPT_ECHO_OFF: /* password */
+                    user_pass = (struct t_user_pass *) appdata_ptr;
+                    reply[i].resp = g_strdup(user_pass->pass);
+                    break;
 
-    for (i = 0; i < num_msg; i++)
+                case PAM_ERROR_MSG:
+                    LOG(LOG_LEVEL_ERROR, "PAM: %s", msg[i]->msg);
+                    break;
```

Comment:
It's not clear to me why this error from pam only gets logged and does not propagate to sesman. If this is a recoverable error, I suggest that the log level be decreased to warning instead of error.

Reply:
That's a really good point which has led me to question a few assumptions I've made here.

This code is reached via callbacks from [pam_conv(3)](https://linux.die.net/man/3/pam_conv). Here's an extract:-

```
       Each message can have one of four types, specified by the msg_style member of struct pam_message:

       PAM_PROMPT_ECHO_OFF
           Obtain a string without echoing any text.

       PAM_PROMPT_ECHO_ON
           Obtain a string whilst echoing text.

       PAM_ERROR_MSG
           Display an error message.

       PAM_TEXT_INFO
           Display some text.
```
I've taken the simplistic approach of using LOG_LEVEL_ERROR for PAM_ERROR_MSG and LOG_LEVEL_INFO for PAM_TEXT_INFO.

My understanding is that the PAM library itself will return an error to sesman, so there's no need to do it here. Your question however prompted me to check that assumption so I had a look at the source for linux-pam.

There's a generic callback implementation here:-

https://github.com/linux-pam/linux-pam/blob/d62120ed7cfa99dec0e66e562f20fb32987c5980/libpam_misc/misc_conv.c#L276-L401

That sends PAM_ERROR_MSG to stderr, and returns PAM_SUCCESS in the absence of any other errors which is what I'm doing above.

The other calls which mention PAM_ERROR_MSG are all returning errors to the application with the exception of one in pam_selinux which allows for a retry before the application is notified.

TLDR; it's pretty certain that logging an error here will be followed by an error being returned to the application.

If we get an error back from PAM sesman will log it with LOG_LEVEL_ERROR at line 234 or 254 in my modified file.

So after some time looking into this, I'm reasonably sure that the behaviour I've implemented is correct but please come back to me if you disagree.

Reply:
I don't know anything about pam, and I've only quickly read some parts of http://www.linux-pam.org/Linux-PAM-html/Linux-PAM_ADG.html

My take away from that guide is that you are correct, and the error will be returned from one of the other pam_* functions.

From my reading it seems that the return value of pam_conv callback represents only the answer to "was the text displayed to the user, or was the input received from the user?" and the error status of the authentication is tracked separately.

Can you please add to the function doc what the return value means, and any relevant references you used in your research.

Reply:
Will do - thanks.